rfcs/rfc-using-global-store-for-form-components.md

```diff
@@ -0,0 +1,317 @@
+- Start Date: 2023-03-06
+- Target Major Version: n/a
+- Reference Issues: bcgov/entity#15509
+- Entity Issue: 
+- Implementation PR: 
+
+# Summary
+
+This RFC seeks feedback to help improve the application architecture of
+business-filings-ui, business-create-ui and business-filings-ui.  At present,
+we use props and events to communicate between form components like this:
+
+```javascript
+<CourtOrderPoa
+  :draftCourtOrderNumber="getCourtOrderStep.courtOrder.fileNumber"
+  :hasDraftPlanOfArrangement="getCourtOrderStep.courtOrder.hasPlanOfArrangement"
+  :courtOrderNumberRequired="false"
+  :invalidSection="isCourtOrderInvalid"
+  @emitCourtNumber="setCourtOrderFileNumber(\$event)"
+  @emitPoa="setHasPlanOfArrangement(\$event)"
+  @emitValid="setCourtOrderValidity(\$event)"
+/>
+```
+
+We're proposing to modify the form components to use Vuex or Pinia getters and mutators instead.
+
+# Basic example
+
+Each form component would be passed a store location as a string like this: 
+```javascript
+<CourtOrderPoa path="some.location.in.the.store" />
+```
+
+The component then gets its default values from the store location and, when the user modifies
+the form, the component mutates the store location provided.
+
+# Motivation
+
+There are three primary goals:
+
+First, separation of concerns.  At present, the parent component must know about each field in 
+the child component.  This makes changes more difficult as both the child and the parent must be
+modified.
+
+Second, make the component reuse easier. Using the new approach, components can be reused calling
+the component and providing a store location.  Using the existing approach, each field's needs to
+be "wired-up" with a prop and an event listener.
+
+Third, make forms that use progressive disclosure easier to implement. Using the existing approach,
+if the child component sometimes returns additional fields, the parent must "know" about the
+fields and handle the additional complexity.  Using the new approach, the knowledge of the 
+additional fields could be contained to the child component making progressive disclosure
+easier to implement.
+
+As side benefit, using the new approach makes the child component fully reactive.  The current 
+approach copies the props to mutable local variables when the component is mounted.  The values
+are not reactive from outside the component once the child component is mounted.
+
+# Detailed design
+
+Here's an example of a simple series of checkboxes:
+
+![img.png](rfc-using-pinia-for-form-components/example-component.png)
+
+This would be implemented by calling 5 checkboxes with the store location (aka: `path`):
+```javascript
+<SingleCheckbox v-for="(item, idx) in checkboxes"
+  :item="item"
+  :id="idx"
+  :key="idx"
+  :path="path"
```

Comment:
okay so now the parent or any other component can look up the value of the checkbox by looking in the store as long as they have the id / path correct?
That is handy. The only thing I don't like very much is that you need to define this in the store for each checkbox list. I guess you could have a common checkboxes list, but you still need that before you can import/use the component

Reply:
>  is that you need to define this in the store for each checkbox list.
I'm not sure I understand.  Can you clarify what would need to be defined in the store?

Reply:
> wouldn't you need to define a module in the store containing the checkbox list before using this component?

No

Reply:
so nothing is defined in the store beforehand? How will I know the types of everything and whats available to me without tracing the code where setAttribute is called?

Reply:
Just look to the store location (aka path + id), that will tell you the type.